contract/src/RadicleCloudBase.sol

```diff
@@ -0,0 +1,272 @@
+        uint64 expiry;
+        uint64 rate;
+    }
+    /// @notice Array of packages that topped up the deployment of org
+    mapping(address => Package[]) public pkgs;
+
+    /// @notice Rate (price) new top ups are charged
+    uint64 public ratePerBlock;
+    /// @notice Rate (price) currently being earned
+    int64 private runningRate = 0;
+
+    /// @notice Duration of a package in terms of blocks
+    uint32 public immutable duration;
+    /// @notice Blocks to wait before being eligible for withdraw
+    uint32 public immutable withdrawWait;
+
+    /// @notice MAXBLOCK is the maximum value of uint64
+    uint64 private constant MAXBLOCK = type(uint64).max;
+
+    /// @dev Add functionalities for index and block retrieval
+    using ImpactPoints for uint64[];
+    /// @notice Special array of blocks when buys happened
+    uint64[] private buys = [1, MAXBLOCK];
+    /// @notice Special array of blocks when expires happen
+    uint64[] private expires = [1, MAXBLOCK];
+    /// @notice Special array of blocks when cancels happened
+    uint64[] private cancels = [1, MAXBLOCK];
+    /// @notice Special array of blocks when renews happened
+    uint64[] private renews = [1, MAXBLOCK];
+    /// @notice Whether a renewal happened at this block
+    mapping(uint64 => bool) private hasRenewal;
+    /// @notice Sum of unaccounted impacts at block t
+    mapping(uint64 => int64) private impacts;
+
+    /// @dev Last block we've withdrawn funds utill.
+    uint64 private lastProcessed;
+
+    /// @notice Emitted when money is deposited, could be renewal or initial order.
+    /// @param org the org owning the deployment
+    /// @param expiry next expiry for this deployment
+    event NewTopUp(address org, uint64 expiry);
+
+    /// @notice Emitted when deployment for org is cancelled or suspended.
+    /// @param org the org owning the deployment
+    event DeploymentStopped(address org);
+
+    /// @notice Modifier to check if caller is owner.
+    modifier isOwner() {
+        require(msg.sender == owner, "Caller is not owner");
+        _;
+    }
+
+    /// @notice Create a new contract.
+    /// @param price amount that should be charged per block
+    /// @param _duration immutable number of blocks each top up lasts
+    /// @param _owner owner of this contract
+    constructor(
+        uint64 price,
+        uint32 _duration,
+        address _owner
+    ) {
+        duration = _duration;
+        withdrawWait = (_duration * 3) / 100;
+        owner = _owner;
+        ratePerBlock = price;
+        lastProcessed = uint64(block.number);
+    }
+
+    /// @notice Change owner of contract.
+    /// @param newOwner address of new owner
+    function changeOwner(address newOwner) public isOwner {
+        owner = newOwner;
+    }
+
+    /// @notice Change ratePerBlock for orders after this block.
+    /// @param newRate the new rate per block
+    function changeRate(uint64 newRate) public isOwner {
+        ratePerBlock = newRate;
```

Comment:
Calling this function will break the contract logic. Take a look at `topUp`, lines 145-147:
```solidity
impacts[dep[org].expiry] += int64(ratePerBlock);
dep[org].expiry = newExpiry;
impacts[dep[org].expiry] -= int64(ratePerBlock);
```
It assumes that `ratePerBlock` doesn't change between calls to `topUp`.

Reply:
I think from a UX side of things, being able to update the rate is not a super important feature. If removing it can simplify the contract a lot, I think it's probably worth it.

Reply:
@CodeSandwich It's working as intended, unless you're talking about an edge case that I'm missing? Let me illustrate:

Suppose you buy a package for `duration = 200` at `t = 0`, then at `t=50` operator doubles `rate => 2` and you renew at `t = 100`, here's the state:

```
impacts		      (+1,)     (-1,+2)      (-2,)   
time			0 ....... 200 ....... 400
======
sum			+1         +1	      -2

revenue (t: 0   -> 200) = 200 x 1
revenue (t: 200 -> 400) = 200 x 2
revenue (t: 400 -> ∞  ) = 200 x 0
```

@cloudhead Operator is dealing with two dynamic variables: 

1. price of cloud resources
2. time it takes for a block to be mined

Thus having a dynamic `rate` is a must, not a luxury.

Reply:
Aah, that's right, this is correct! After the rate change we can ignore existing impacts and keep appending new ones. We can also cancel, because the whole history is stored in `pkgs`.

So the behavior of `changeRate` is that existing subscriptions are unchanged, but future `topUp`s are going to buy a different amount of time. You've probably already analyzed this and it's fine, but is this rigidness good for the ecosystem? If I subscribe for 5 years, I'm forcing the operator to bet on the future computation price. If it goes down, I can cancel and topUp again, but if it goes up, the operator can't do anything except silently refusing to serve me, but then they'll probably pay some penalties. On the other hand if the rate could be changed immediately, the operator could steal all the funds in the contract with a single call.

Reply:
We probably have to limit number of outstanding packages to ideally no more than a year, but if someone orders 5 year in advanced, well, then they deserve to be de-risked, and operator has to hedge their own side. This, in addition to penalty system, should encourage honest behavior, I think.

Reply:
The point is that the user ordering a long period isn't risking anything. If they don't like the arrangement, they can cancel any time and if at any moment in the future there's a better rate, they can cancel and topUp to use it. It's hard for the operator to increase the price, because many people will stick with the old one, but if the operator lowers the price, they can be sure that everybody will switch to it almost immediately. It may be a desirable property, I don't know.

Reply:
They might not be risking per se, but they are committing, the fact that they are parking X amount here, has opportunity cost for them, so it's not entirely free. Also, as of now, if they cancel, they'll end up losing part of their data, which they might never recover. It's not completely free.

An smarter approach would be for orgs to watch rate changes in mempool, and front-run it by topping up. Again, this can be mitigated by a limit on max length of package.